### DISCUSSION TITLE
viem/op-stack Prove/Finalize transaction is failing due to conflict between wallet extensions

### DISCUSSION THREAD

**Discussion by danielsimao** [wevm/viem#2403] [Question]
### Check existing issues

- [X] I checked there isn't [already an issue](https://github.com/wevm/viem/issues) for the bug I encountered.

### Viem Version

2.10.2

### Current Behavior

I have metamask and okx wallet extension. If I connect metamask and I try to prove or finalize a withdraw, I get the following error: 

Prove:  TransactionExecutionError
    at ut (getTransactionError.js:11:12)
    at lt (sendTransaction.js:126:15)
    at async Object.mutationFn (WithdrawStatus.tsx:55:20)Caused by: ChainMismatchError
    at ct (assertCurrentChain.js:6:15)
    at lt (sendTransaction.js:69:13)
    at async Object.mutationFn (WithdrawStatus.tsx:55:20)

There is no signing prompt being shown.
    

### Expected Behavior

I expect transaction to be proven or finalized without having to disable OKX wallet.

### Steps To Reproduce

1. Connect Metamask (have OKX wallet installed)
2. Prove tx
3. See error mentioned above

### Link to Minimal Reproducible Example

It is a bit hard to create an environment for this, but simply by using the code in the viem/op-stack, you should be able to reach this error.

**Comment by jxom** [comment]
Looks like you are encountering a `ChainMismatchError`. Are you sure you are executing proveWithdrawal on the right chain?

**Comment by danielsimao** [comment]
@jxom Thank you for such a quick response.

Yes, the error only occurs when metamask is connect and OKX is enabled as an extension. If I disabled OKX wallet, the tx goes through.

This is where the error occurs: https://github.com/bob-collective/bob-ui/blob/main/apps/evm/src/pages/Bridge/components/BridgeStatus/WithdrawStatus.tsx#L54

I do check for chain correctness outside of this code, so I do confirm that this error still occurs even on the correct chain (L1).

**Comment by danielsimao** [comment]
I can share a recording of the app through discord if you need.

**Comment by jxom** [comment]
I don't really think this is an issue in Viem, but rather upstream in wallet orchestration logic. I cannot seem to view your repo as it is private.

Are you using `window.ethereum` as your wallet EIP-1193 provider? I would strongly recommend not to rely on `window.ethereum` as conflicts like these can happen. 

I would strongly recommend either:
- High-level: Using [Wagmi](https://wagmi.sh) with [`usePublicClient`](https://wagmi.sh/react/api/hooks/usePublicClient#usepublicclient) & [`useWalletClient`](https://wagmi.sh/react/api/hooks/useWalletClient), or
- Low-level: Extract the provider from [EIP-6963 events](https://eips.ethereum.org/EIPS/eip-6963) via [mipd](https://github.com/wevm/mipd)

**Reply by danielsimao** [reply]
Sorry @jxom, I shared the wrong link. Could you take a look at this one?

https://github.com/bob-collective/ui/blob/main/apps/evm/src/pages/Bridge/components/BridgeStatus/WithdrawStatus.tsx#L55

I follow 100% what you guys have on your docs and yes I'm using the hooks provided by wagmi.

So, what you are saying is that this goes down to an issue with OKX wallet right?

Once again, thank you for taking the time to help me. I know your team is pretty busy.

In case you have time, could you give some feedback on this: https://github.com/wevm/viem/discussions/2250? I raised this also with optimism. 

Cheers
